contracts/OwnedSet.sol

```diff
@@ -56,85 +70,128 @@ contract OwnedSet is Owned, ValidatorSet {
-	address constant SYSTEM_ADDRESS = 0xffffFFFfFFffffffffffffffFfFFFfffFFFfFFfE;
-	uint public recentBlocks = 20;
-
-	// Current list of addresses entitled to participate in the consensus.
-	address[] validators;
-	address[] pending;
-	mapping(address => AddressStatus) pendingStatus;
-	// Was the last validator change finalized. Implies validators == pending
-	bool public finalized;
-
-	function OwnedSet(address[] _initial) public {
+	constructor(address[] _initial)
+		public
+	{
 		pending = _initial;
-		for (uint i = 0; i < _initial.length - 1; i++) {
+		for (uint i = 0; i < _initial.length; i++) {
 			pendingStatus[_initial[i]].isIn = true;
 			pendingStatus[_initial[i]].index = i;
 		}
 		validators = pending;
 	}
 
-	// Called to determine the current set of validators.
-	function getValidators() constant public returns (address[]) {
-		return validators;
-	}
-
-	function getPending() constant public returns (address[]) {
-		return pending;
-	}
-
-	function finalizeChange() public onlySystemAndNotFinalized {
-		validators = pending;
-		finalized = true;
-		emit ChangeFinalized(getValidators());
+	// Called when an initiated change reaches finality and is activated.
+	function finalizeChange()
+		external
+		onlySystemAndNotFinalized
+	{
+		finalizeChangeInternal();
 	}
 
 	// OWNER FUNCTIONS
 
 	// Add a validator.
-	function addValidator(address _validator) public onlyOwner isNotPending(_validator) {
+	function addValidator(address _validator)
+		external
+		onlyOwner
+		isNotPending(_validator)
+	{
 		pendingStatus[_validator].isIn = true;
 		pendingStatus[_validator].index = pending.length;
 		pending.push(_validator);
 		initiateChange();
 	}
 
 	// Remove a validator.
-	function removeValidator(address _validator) public onlyOwner isPending(_validator) {
-		pending[pendingStatus[_validator].index] = pending[pending.length - 1];
+	function removeValidator(address _validator)
+		external
+		onlyOwner
+		isPending(_validator)
+	{
+		// Remove validator from pending by moving the
+		// last element to its slot
+		uint index = pendingStatus[_validator].index;
+		pending[index] = pending[pending.length - 1];
+		pendingStatus[pending[index]].index = index;
 		delete pending[pending.length - 1];
 		pending.length--;
-		// Reset address status.
+
+		// Reset address status
 		delete pendingStatus[_validator].index;
 		pendingStatus[_validator].isIn = false;
+
 		initiateChange();
 	}
 
-	function setRecentBlocks(uint _recentBlocks) public onlyOwner {
+	function setRecentBlocks(uint _recentBlocks)
+		external
+		onlyOwner
+	{
 		recentBlocks = _recentBlocks;
 	}
 
 	// MISBEHAVIOUR HANDLING
 
-	// Called when a validator should be removed.
-	function reportMalicious(address _validator, uint _blockNumber, bytes /* _proof */) public onlyOwner isRecent(_blockNumber) {
+	// Report that a validator has misbehaved maliciously.
+	function reportMalicious(address _validator, uint _blockNumber, bytes _proof)
+		external
+		onlyOwner
```

Comment:
I don't think this should be restricted to the `owner`. I think at least any validator should be able to report misbehavior.